Background: add background attachment support to theme.json styles

ramonjd opened the pull request:
Syncing changes from https://github.com/WordPress/gutenberg/pull/61382

> [!NOTE]
> This PR should be committed after https://github.com/WordPress/wordpress-develop/pull/6836 to ensure a clean rebase 😄 

## Testing

Unit tests:

```
npm run test:php -- --filter Tests_Theme_wpThemeJson && npm run test:php -- --filter Tests_Block_Supports_WpRenderBackgroundSupport && npm run test:php -- --filter Tests_wpStyleEngine
```

This PR adds the backend functionality only, which means the editor UI isn't yet available.

To test, define a background image for the site in theme.json. 

Here's an example using an image in src/wp-content/themes/twentytwentyfour/

```json
{
	"$schema": "../../schemas/json/theme.json",
	"version": 3,
	"settings": {
		"appearanceTools": true
	},
	"styles": {
		"background": {
				"backgroundImage": {
					"url": "file:./assets/images/art-gallery.webp"
				},
				"backgroundAttachment": "fixed"
		},
		"blocks": {
			"core/verse": {
				"background": {
					"backgroundImage": {
						"url": "file:./assets/images/abstract-geometric-art.webp"
					}
				}
			}
        }
   	}
}
```

Now check the site:


https://github.com/user-attachments/assets/6679d670-b306-45b4-858a-0b490bee5595


Trac ticket: https://core.trac.wordpress.org/ticket/61720

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

ramonjd committed:
Initial commit - bringing across changes from https://github.com/WordPress/gutenberg/pull/61382

ramonjd committed:
Tabs. No spaces.

ramonjd committed:
update style engine

ramonjd committed:
Complete theme.json tests and add ticket number

mukeshpanchal27 commented on src/wp-includes/class-wp-theme-json.php:
```diff
@@ -237,6 +237,7 @@ class WP_Theme_JSON {
 		'background-position'               => array( 'background', 'backgroundPosition' ),
 		'background-repeat'                 => array( 'background', 'backgroundRepeat' ),
 		'background-size'                   => array( 'background', 'backgroundSize' ),
+		'background-attachment'             => array( 'background', 'backgroundAttachment' ),
```
Add the property in docblock with `@since`

ramonjd committed:
Added since attributes

andrewserong commented on tests/phpunit/tests/style-engine/styleEngine.php:
```diff
@@ -539,22 +539,24 @@ public function data_wp_style_engine_get_styles() {
 				'block_styles'    => array(
 					'background' => array(
-						'backgroundImage'    => array(
+						'backgroundImage'      => array(
 							'url' => 'https://example.com/image.jpg',
 						),
-						'backgroundPosition' => 'center',
-						'backgroundRepeat'   => 'no-repeat',
-						'backgroundSize'     => 'cover',
+						'backgroundPosition'   => 'center',
+						'backgroundRepeat'     => 'no-repeat',
+						'backgroundSize'       => 'cover',
+						'backgroundAttachment' => 'fixed',
```
Tiny nit: for these changes, do we also need to add the ticket number to the above `test_wp_style_engine_get_styles` test?

ramonjd replied:
Whoops, missed that. Thanks for catching it. 

Done.

andrewserong reviewed (approved):
This code change looks good to me and is testing nicely in the site frontend 👍 

For anyone else reviewing, some of the PHP changes in `block-supports/background.php` also include changes from https://github.com/WordPress/wordpress-develop/pull/6836, as mentioned in the PR description, so it looks like that PR should land first. Factoring in #6836, the changes here match the expected changes from https://github.com/WordPress/gutenberg/pull/61382 👍

LGTM!

ramonjd committed:
Add ticket annotation

mukeshpanchal27 commented on tests/phpunit/tests/theme/wpThemeJson.php:
```diff
@@ -5020,24 +5022,24 @@ public function test_get_top_level_background_image_styles() {
 			)
 		);
 
-		$expected_styles = "html{min-height: calc(100% - var(--wp-admin--admin-bar--height, 0px));}:root :where(body){background-image: url('http://example.org/image.png');background-position: center center;background-repeat: no-repeat;background-size: contain;}";
```
Nice!

ramonjd committed:
Add block style tests

ramonjd commented:
I've rebased on top of https://github.com/WordPress/wordpress-develop/pull/6836 and have added theme json block-level tests.

One thing to note is that https://github.com/WordPress/wordpress-develop/pull/6836 opens up block-level global styles, so I've updated the theme.json test snippet in the description.

🍺

noisysocks commented:
@ramonjd: Is this ready to commit?

ramonjd commented:
> Is this ready to commit?

Oh yes, please!!

noisysocks commented:
I intended that as a nudge for you... 😂 You win this round, I was unclear.

ramonjd commented:
> I intended that as a nudge for you... 😂 You win this round, I was unclear.

Haha. No worries. I mistakenly assumed there was some rule about committing your own patches.

ramonjd commented:
Merged in [[58834]](https://core.trac.wordpress.org/changeset/58834)

https://github.com/WordPress/wordpress-develop/commit/c3e1d3a8b8c2393ce8d6d50cededbea479516b8e

noisysocks commented:
Thanks!
